### PR TITLE
Ensure that stdin/stdout/stderr are not mapped 1:1 under NODERAWFS

### PR DESCRIPTION
**Pull request by kleisauke**
Previously, when linking with -sNODERAWFS, the stdin, stdout and stderr streams were respectively attached to file descriptors 0, 1 and 2 of the running Node process.
https://github.com/emscripten-core/emscripten/blob/52e380517269b9e2d9341ff80f54e0b6f7398898/src/library_noderawfs.js#L47-L52

This implicitly means that overriding the print, printErr, stdin, stdout and stderr handlers on the incoming module wasn't possible.
https://github.com/emscripten-core/emscripten/blob/52e380517269b9e2d9341ff80f54e0b6f7398898/src/shell.js#L431-L432
https://github.com/emscripten-core/emscripten/blob/52e380517269b9e2d9341ff80f54e0b6f7398898/src/library_fs.js#L1496-L1499

This commit ensures that stdin/stdout/stderr uses the library_tty.js implementation for its printing, which leverages the console object, whenever the above handlers weren't overriden. This matches what is done when filesystem support isn't included.

Resolves: #17688.
Resolves: #22264.

### CODE REVIEW

**Comment by sbc100** [comment]
Is this really the behaviour we want though?  Doesn't NODERAWFS imply that you do want write directly the stdin/stdout of the node process?

**Comment by kleisauke** [comment]
Good point, it looks like this was included in the initial implementation (see #6008), so perhaps there should be some discussion about it first.

Writing directly to `stdout`/`stderr` is sometimes not desirable, at least from a library point of view. The behavior is still the same if these handlers on the incoming module are not overridden. For example, `console.warn` will output to `stderr` by default and `stdin` usage is still working on this branch.
```bash
$ node -v
v14.18.2
$ node -e "console.warn('hello, world\!')" > /dev/null
hello, world!
$ emcc test/module/test_stdin.c -sNODERAWFS -sFORCE_FILESYSTEM -sEXIT_RUNTIME
$ echo 'hello, world!' | node a.out.js
hello, world!
eof
```

This PR also unifies its behavior with the other file systems (e.g. WasmFS also uses `out()` and `err()` for printing on the Node backend, see [`system/lib/wasmfs/special_files.cpp`](https://github.com/emscripten-core/emscripten/blob/29f65fa7babef81cc00dfe9ce8d36de482247a70/system/lib/wasmfs/special_files.cpp)) and when file system support is not needed/detected (the mentioned issue occurred only with `-sFORCE_FILESYSTEM` for a simple `printf()` program).

**Review by sbc100 — APPROVED** [review]

**Comment by kleisauke** [comment]
I'm marking this as a draft, since I'm a bit worried that this may cause performance issues like issue https://github.com/emscripten-core/emscripten/issues/16755, which doesn't affect the `NODERAWFS` setting currently.

IIUC, pyodide currently resolves(?) that issue within the above linked PR (/cc @hoodmane).

**Comment by kleisauke** [comment]
I've marked this PR as ready for review. I think issue #16755 should be addressed separately and not block this one.

File: test/test_other.py [hunk 1]
```diff
@@ -13476,6 +13477,38 @@ def test_unistd_close_noderawfs(self):
...
+    self.do_runf(test_file('hello_world.c'))
+
+  @requires_node
+  def test_noderawfs_override_stdin(self):
+    self.set_setting('NODERAWFS')
+    self.set_setting('FORCE_FILESYSTEM')
+    self.set_setting('EXIT_RUNTIME')
+    create_file('pre.js', '''
+const data = 'hello, world!\\n'.split('').map(c => c.charCodeAt(0));
+Module['stdin'] = () => data.shift() || null;
```
Comment (sbc100) [inline test/test_other.py:13507]: It looks like we don't have any existing tests for `Module['stdin'`].   Can you perhaps at this new test here can call it `test_module_stdin`?  And then annotate it with `also_with_noderawfs` (and `@crossplatform`).

Comment (kleisauke) [reply test/test_other.py]: Addressed with commit f927a004df016ad234b65e4f4b130f22a3044626.

File: test/unistd/isatty.c [hunk 2]
```diff
@@ -21,7 +21,11 @@ int main() {
   assert(errno == EBADF);
 
   err = isatty(open("/dev/stdin", O_RDONLY));
+#ifdef NODERAWFS
+  assert(!err);
+#else
```
Comment (sbc100) [inline test/unistd/isatty.c:26]: Why is this needed? Perhaps add a comment?

File: test/test_other.py [hunk 3]
```diff
@@ -1689,6 +1689,18 @@ def test_stdin(self, args):
...
+  @crossplatform
+  def test_module_stdin(self):
+    self.set_setting('FORCE_FILESYSTEM')
+    self.set_setting('EXIT_RUNTIME')
+    create_file('pre.js', '''
+const data = 'hello, world!\\n'.split('').map(c => c.charCodeAt(0));
+Module['stdin'] = () => data.shift() || null;
+''')
+    self.emcc_args += ['--pre-js', 'pre.js']
+    self.do_runf(test_file('module/test_stdin.c'), 'hello, world!')
```
Comment (sbc100) [inline test/test_other.py:1702]: No need for `test_file` here.

File: test/unistd/isatty.c (hunk 2 above)
Comment (kleisauke) [reply test/unistd/isatty.c]: It was intended as a regression test, but it appears that `other.test_unistd_isatty` was already functioning correctly with `NODERAWFS`. Therefore, I reverted this with commit c39e5b14d3510f4fa673feeec3c76a85664a645f.

File: test/test_other.py (hunk 3 above)
Comment (kleisauke) [reply test/test_other.py]: Ah, great, done with commit c39e5b14d3510f4fa673feeec3c76a85664a645f.

File: test/test_other.py [hunk 4]
```diff
@@ -13492,6 +13505,26 @@ def test_unistd_close_noderawfs(self):
     self.emcc_args += ['--pre-js', 'pre.js']
     self.do_run_in_out_file_test('unistd/close.c')
 
+  @requires_node
+  def test_noderawfs_override_standard_streams(self):
```
Comment (sbc100) [inline test/test_other.py:13509]: Should we call this test `test_module_stdout_stderr` and run it both with and without wasmfs?  (and move it up to be alongside `test_module_stdin`?

Comment (kleisauke) [reply test/test_other.py]: Good idea, done with commit 02a4c50e247b4359a941e65d5108492b20b8afc7.

**Review by sbc100 — APPROVED** [review]
Landing as is since the browser test failures are unrelated (and fixed on main already)

**Comment by sbc100** [comment]
It looks like this is breaking on windows sadly.  I guess we don't have any noderawfs tests tagged as @crossplatform so it wasn't getting tested in github CI on windows :(

On the emscripten-releases roller I'm seeing lots of failures: https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8740021917817007009/+/u/Emscripten_testsuite__other_/stdout

The all looks like this:

```
ErrnoError: No such file or directory
    at Object.open (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:3282:17)
    at Object.createStandardStreams (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:3615:24)
    at Object.init (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:3651:12)
    at Object.init (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:2199:13)
    at initRuntime (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:420:6)
    at doRun (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:4785:5)
    at run (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:4806:5)
    at runCaller (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:4726:19)
    at removeRunDependency (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:552:7)
    at receiveInstance (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:736:5)
    at receiveInstantiationResult (C:\b\s\w\ir\x\t\emtest_jyle_eyr\emscripten_test_other_i3dmf5t9\a.out.js:754:5) {
  errno: 44,
  code: 'ENOENT'
```
  
Which means I think the issue is with createStandardStreams, which is no longer overridden here:

https://github.com/emscripten-core/emscripten/blob/bd8b894fc7ad236aa232db04026a7248bc5c9035/src/library_fs.js#L1431-L1434

**Comment by sbc100** [comment]
I think we should perhaps revert, and at the same time add `@crossplatform` to `test_noderawfs` in `test_other.py`, then we can attempt to re-land with windows testing?

**Comment by dschuff** [comment]
This is also breaking the `ray` test from the LLVM testsuite (the output is [here](https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8739594615239511041/+/u/LLVM_testsuite/stdout) FTR but honestly don't bother as that log isn't useful).
That test ([source](https://chromium.googlesource.com/native_client/pnacl-llvm-testsuite/+/refs/heads/main/SingleSource/Benchmarks/Misc-C++/Large/ray.cpp) seems to just use `cout` to output text and then what appears to be binary data casted to char. I haven't looked in detail at what's happening yet, other than to verify that the output is indeed different.

**Comment by dschuff** [comment]
With a very small reproducer like 
```
int main(void) {
  std::cout << "Some ASCII text\n";
  for (int i = 0; i < 256; ++i)
    std::cout << char(i);
  return 0;
}
```
You can see differences in outputting 0-valued bytes, and when the counter rolls from 0x7f to 0x80, i.e. out of the ASCII range. So presumably because stdout is open in text mode, there may be a difference in how that gets handled, relative to what the test expects (and it's possible that it's relying on something it shouldn't). But I would still like to understand what's happening here, e.g. is this a behavior difference in the emscripten runtime or node, etc.

**Comment by kleisauke** [comment]
@dschuff Sorry for the breakage, looks like `NODERAWFS` is lacking some test coverage. I just opened PR #22393 for this.

> You can see differences in outputting 0-valued bytes, and when the counter rolls from 0x7f to 0x80, i.e. out of the ASCII range.

Hmm, I was also able to reproduce this without `NODERAWFS`, it seems to trigger this warning:
https://github.com/emscripten-core/emscripten/blob/eb4d36d47f451d0040a0c796606eaae9aa530de3/src/library_strings.js#L81

**Comment by kleisauke** [comment]
FWIW, here's a standalone reproducer for that ASCII issue:

Expected behavior:
```console
$ curl -LO https://github.com/llvm/llvm-test-suite/raw/main/SingleSource/Benchmarks/Misc-C++/Large/ray.cpp
$ em++ -O3 -sEXIT_RUNTIME ray.cpp -o ray.js
$ node ray.js > x.ppm
$ sha256sum x.ppm
cd6fa6fa75b4ba29077f7e4e531b96f17ff0b82f3c70df526c36f748d6a38c5c  x.ppm
$ vips copy x.ppm x.png
```
<details>
  <summary>x.png</summary>

x
</details>

Current/actual behavior:
```console
$ em++ -O3 -sEXIT_RUNTIME ray.cpp -o ray.js
$ node ray.js > x.ppm
$ sha256sum x.ppm
acaa1db08ee30dafda03b41c77466b1435fe8ffdd4d88e5a5876e87dff126f58  x.ppm
$ vips copy x.ppm x.png
VipsImage: memory area too small --- should be 262144 bytes, you passed 79485
```

**Comment by kleisauke** [comment]
... I just revised PR #22393 to avoid `library_tty.js` for stdio streams when linking with `-sNODERAWFS`, which should fix this particular `ray` test.

Of course, the bug still occurs without `-sNODERAWFS` (i.e. when `library_tty.js` is used), but let's address that separately.
